src/selectors/ContainerSelector.h

```diff
@@ -0,0 +1,129 @@
+class ContainerSelector {
+ public:
+  /**
+   * Constructor for the ContainerSelecor class.
+   * @param boxMin Lower corner of the container.
+   * @param boxMax Upper corner of the container.
+   * @param cutoff  Cutoff radius to be used in this container.
+   * @param allowedContainerOptions Vector of container types the selector can choose from.
+   * @param allowedTraversalOptions Vector of traversals the selector can choose from.
+   */
+  ContainerSelector(std::array<double, 3> &boxMin, std::array<double, 3> &boxMax, double cutoff,
+                    std::vector<ContainerOptions> &allowedContainerOptions,
+                    std::vector<TraversalOptions> &allowedTraversalOptions)
+      : _boxMin(boxMin),
+        _boxMax(boxMax),
+        _cutoff(cutoff),
+        _allowedContainerOptions(allowedContainerOptions),
+        _allowedTraversalOptions(allowedTraversalOptions),
+        _optimalContainer(nullptr) {}
+
+  /**
+   * Getter for the optimal container.
+   * @return Smartpointer to the optimal container.
+   */
+  std::shared_ptr<ParticleContainer<Particle, ParticleCell>> getOptimalContainer();
+
+  /**
+   * Evaluates the optimal container option.
+   */
+  void tune();
+
+ private:
+  std::vector<std::unique_ptr<ParticleContainer<Particle, ParticleCell>>> generateContainers();
+  void chooseOptimalContainer(std::vector<std::unique_ptr<ParticleContainer<Particle, ParticleCell>>> containers);
+
+  std::array<double, 3> _boxMin, _boxMax;
+  double _cutoff;
+  std::vector<ContainerOptions> _allowedContainerOptions;
+  std::vector<TraversalOptions> _allowedTraversalOptions;
+  std::shared_ptr<ParticleContainer<Particle, ParticleCell>> _optimalContainer;
+};
+
+template <class Particle, class ParticleCell>
+std::vector<std::unique_ptr<ParticleContainer<Particle, ParticleCell>>>
+ContainerSelector<Particle, ParticleCell>::generateContainers() {
+  std::vector<std::unique_ptr<ParticleContainer<Particle, ParticleCell>>> containers;
+
+  for (auto &option : _allowedContainerOptions) {
+    switch (option) {
+      case directSum: {
+        containers.push_back(std::make_unique<DirectSum<Particle, ParticleCell>>(_boxMin, _boxMax, _cutoff));
+        break;
+      }
+      case linkedCells: {
+        containers.push_back(
+            std::make_unique<LinkedCells<Particle, ParticleCell>>(_boxMin, _boxMax, _cutoff, _allowedTraversalOptions));
+        break;
+      }
+      case verletLists: {
+        containers.push_back(std::make_unique<DirectSum<Particle, ParticleCell>>(_boxMin, _boxMax, _cutoff));
+        break;
+      }
+      default: { AutoPasLogger->warn("Container type {} is not a known type!", option); }
+    }
+  }
+
+  assert(containers.size() > 0);
+  return containers;
+}
+
+template <class Particle, class ParticleCell>
+void ContainerSelector<Particle, ParticleCell>::chooseOptimalContainer(
+    std::vector<std::unique_ptr<ParticleContainer<Particle, ParticleCell>>> containers) {
+  // TODO: Autotuning goes here
```

Comment:
One could introduce an extra parameter for selecting a tuner or apply the strategy pattern again and pass a pointer to a tuner.
(And then do tuning over the tuners :D)

Reply:
In particular, I think it would make sense to create the CellFunctor here and then call iteratePairwise with it. This would remove the need for an AoS and SoA version. Also, a selector/tuner for Newton3 and Data structure could then be plugged in here. The cell functor could then be thrown at a decision tree like you suggested in #52 .
@SteffenSeckler opinions?